<img align="right" style="padding-left:10px; height: 20%; width: 20%" src="figures/projector-300x300.jpg" ></a>

## Case Study: Movie Suggestion

### The Movies Dataset

Collaborative datasets for movies (and other products) can be large! Here is a [small (1 MB) subset of the IMDB database](https://grouplens.org/datasets/movielens/latest/), downloaded and unzipped for your convenience.

The dataset consists of 9742 movies.

In [37]:
import pandas as pd
import numpy as np
from pandas import DataFrame
movies_directory = '../../04-analysis-and-visualization/04-05-recommendations/ml-latest-small/'
movies = pd.read_csv(movies_directory+'movies.csv',header = 0) 

movies

movieId                                              title  \
0           1                                   Toy Story (1995)   
1           2                                     Jumanji (1995)   
2           3                            Grumpier Old Men (1995)   
3           4                           Waiting to Exhale (1995)   
4           5                 Father of the Bride Part II (1995)   
5           6                                        Heat (1995)   
6           7                                     Sabrina (1995)   
7           8                                Tom and Huck (1995)   
8           9                                Sudden Death (1995)   
9          10                                   GoldenEye (1995)   
10         11                     American President, The (1995)   
11         12                 Dracula: Dead and Loving It (1995)   
12         13                                       Balto (1995)   
13         14                                       Nixon (1995)   
14         15                            Cutthroat Island (1995)   
15         16                                      Casino (1995)   
16         17                       Sense and Sensibility (1995)   
17         18                                  Four Rooms (1995)   
18         19              Ace Ventura: When Nature Calls (1995)   
19         20                                 Money Train (1995)   
20         21                                  Get Shorty (1995)   
21         22                                     Copycat (1995)   
22         23                                   Assassins (1995)   
23         24                                      Powder (1995)   
24         25                           Leaving Las Vegas (1995)   
25         26                                     Othello (1995)   
26         27                                Now and Then (1995)   
27         28                                  Persuasion (1995)   
28         29  City of Lost Children, The (Cité des enfants p...   
29         30  Shanghai Triad (Yao a yao yao dao waipo qiao) ...   
...       ...                                                ...   
9712   188189                         Sorry to Bother You (2018)   
9713   188301                        Ant-Man and the Wasp (2018)   
9714   188675                                      Dogman (2018)   
9715   188751                Mamma Mia: Here We Go Again! (2018)   
9716   188797                                         Tag (2018)   
9717   188833              The Man Who Killed Don Quixote (2018)   
9718   189043                                  Boundaries (2018)   
9719   189111                                      Spiral (2018)   
9720   189333               Mission: Impossible - Fallout (2018)   
9721   189381                                    SuperFly (2018)   
9722   189547                                Iron Soldier (2010)   
9723   189713                              BlacKkKlansman (2018)   
9724   190183                           The Darkest Minds (2018)   
9725   190207                                        Tilt (2011)   
9726   190209                 Jeff Ross Roasts the Border (2017)   
9727   190213                                   John From (2015)   
9728   190215                                Liquid Truth (2017)   
9729   190219                                       Bunny (1998)   
9730   190221  Hommage à Zgougou (et salut à Sabine Mamou) (2...   
9731   191005                                     Gintama (2017)   
9732   193565                          Gintama: The Movie (2010)   
9733   193567  anohana: The Flower We Saw That Day - The Movi...   
9734   193571                                Silver Spoon (2014)   
9735   193573            Love Live! The School Idol Movie (2015)   
9736   193579           Jon Stewart Has Left the Building (2015)   
9737   193581          Black Butler: Book of the Atlantic (2017)   
9738   193583                       No Game No Life: Zero (201

### Movie Ratings

The above 9742 movies were rated by 610 users; this works out to about 165 movies on average rated by each user, available in the `ratings.csv` file as sampled in the DataFrame below.

In [38]:
ratings = pd.read_csv(movies_directory+'ratings.csv',header = 0) 
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
5            1       70     3.0   964982400
6            1      101     5.0   964980868
7            1      110     4.0   964982176
8            1      151     5.0   964984041
9            1      157     5.0   964984100
10           1      163     5.0   964983650
11           1      216     5.0   964981208
12           1      223     3.0   964980985
13           1      231     5.0   964981179
14           1      235     4.0   964980908
15           1      260     5.0   964981680
16           1      296     3.0   964982967
17           1      316     3.0   964982310
18           1      333     5.0   964981179
19           1      349     4.0   964982563
20           1      356     4.0   964980962
21           1      362     5.0   964982588
22           1      367     4.0   964981710
23           1      423     3.0   964982363
24           1      441     4.0   964980868
25           1      457     5.0   964981909
26           1      480     4.0   964982346
27           1      500     3.0   964981208
28           1      527     5.0   964984002
29           1      543     4.0   964981179
...        ...      ...     ...         ...
100806     610   150401     3.0  1479543210
100807     610   152077     4.0  1493845817
100808     610   152081     4.0  1493846503
100809     610   152372     3.5  1493848841
100810     610   155064     3.5  1493848456
100811     610   156371     5.0  1479542831
100812     610   156726     4.5  1493848444
100813     610   157296     4.0  1493846563
100814     610   158238     5.0  1479545219
100815     610   158721     3.5  1479542491
100816     610   158872     3.5  1493848024
100817     610   158956     3.0  1493848947
100818     610   159093     3.0  1493847704
100819     610   160080     3.0  1493848031
100820     610   160341     2.5  1479545749
100821     610   160527     4.5  1479544998
100822     610   160571     3.0  1493848537
100823     610   160836     3.0  1493844794
100824     610   161582     4.0  1493847759
100825     610   161634     4.0  1493848362
100826     610   162350     3.5  1493849971
100827     610   163937     3.5  1493848789
100828     610   163981     3.5  1493850155
100829     610   164179     5.0  1493845631
100830     610   166528     4.0  1493879365
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

### Investigating the ratings dataset

Find the number of ratings created by each user.

In [39]:
# Ratings Group By userId:
# print (type(ratings.groupby(["userId"])["userId"].count())) # prints <class 'pandas.core.series.Series'>

# Convert Series to DataFrame
#     Ref: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.reset_index.html#pandas.Series.reset_index

counts = ratings.groupby(["userId"])["userId"].count().reset_index(name="Count")
counts

userId  Count
0         1    232
1         2     29
2         3     39
3         4    216
4         5     44
5         6    314
6         7    152
7         8     47
8         9     46
9        10    140
10       11     64
11       12     32
12       13     31
13       14     48
14       15    135
15       16     98
16       17    105
17       18    502
18       19    703
19       20    242
20       21    443
21       22    119
22       23    121
23       24    110
24       25     26
25       26     21
26       27    135
27       28    570
28       29     81
29       30     34
..      ...    ...
580     581     40
581     582     56
582     583     56
583     584     83
584     585     61
585     586    208
586     587    165
587     588     56
588     589     40
589     590    728
590     591     54
591     592     94
592     593    103
593     594    232
594     595     20
595     596    411
596     597    443
597     598     21
598     599   2478
599     600    763
600     601    101
601     602    135
602     603    943
603     604    100
604     605    221
605     606   1115
606     607    187
607     608    831
608     609     37
609     610   1302

[610 rows x 2 columns]

<a href="https://en.wikipedia.org/wiki/Collaborative_filtering"><img align="right" style="padding-left:10px; height: 40%; width: 40%" src="https://upload.wikimedia.org/wikipedia/commons/5/52/Collaborative_filtering.gif" ></a>

## General Approach

As discussed in [04-05-recommendations](../../04-analysis-and-visualization/04-05-recommendations/04-05-recommendations.ipynb) a generalized version of Collaborative filtering, implied by the adjoining image, is a three-step process:

1. A user expresses their preferences by rating items (e.g. books, movies or CDs) of the system. These ratings can be viewed as an approximate representation of the user's interest in the corresponding domain. _The ratings have been collected by IMDB and imported into the `ratings` DataFrame._
2. The system matches this user's ratings against other users' and finds the people with most "similar" tastes. For the purpose of this Case, we shall determine the recommendations for the user with **userId = 607**.
3. With similar users, the system recommends items that the similar users have rated highly but not yet being rated by this user (presumably the absence of rating is often considered as the unfamiliarity of an item).

<span style="color:blue">

### Solution Development
</span>

We proceed with the calculations as outlined above but first create a **tiny dataset** such that we can develop the solution and _verify the calculations manually._ 

Step 2 of the algorithms is for the system to match this user's ratings against other users' and finds the people with most "similar" tastes. We shall use **userId = 5** for this tiny dataset.

Once the solution has been developed, we will write functions and classes to package the developed code and use it for the given dataset.

In [40]:
# Cell 4
# Initial Parameters
given_userId = 5
threshold_distance = 3.5

In [41]:
# Cell 5
# Data Location
tiny_directory = '../../04-analysis-and-visualization/04-05-recommendations/ml-latest-tiny/'
tiny_movies = pd.read_csv(tiny_directory+'movies.csv',header = 0) 
tiny_movies

movieId                     title  \
0        1          Toy Story (1995)   
1        2            Jumanji (1995)   
2        3   Grumpier Old Men (1995)   
3        4  Waiting to Exhale (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance

In [42]:
# Cell 6
# Drop the timestamp column because we shall not be using it.
tiny_ratings = pd.read_csv(tiny_directory+'ratings.csv',header = 0).drop(columns=['timestamp'])
tiny_ratings

userId  movieId  rating
0       1        1     4.0
1       1        3     4.0
2       5        1     4.0
3       6        2     4.0
4       6        3     5.0
5       6        4     3.0
6       7        1     4.5
7       8        2     4.0
8      15        1     2.5

### Librares to use

We will be using numpy and scipy for most of the calculations, mostly use `pandas` for pretty printing. 

Variable names will be chosen such that:

1. Variables ending in `_np` will be used for numpy arrays.
1. Variables ending in `_2d` will be used for 2-D numpy arrays.
2. Variables ending in `_df` will be used for pandas DataFrames.

In [43]:
# Cell 7
tiny_ratings_np = tiny_ratings.to_numpy(dtype=np.float32)
tiny_ratings_np

array([[ 1. ,  1. ,  4. ],
       [ 1. ,  3. ,  4. ],
       [ 5. ,  1. ,  4. ],
       [ 6. ,  2. ,  4. ],
       [ 6. ,  3. ,  5. ],
       [ 6. ,  4. ,  3. ],
       [ 7. ,  1. ,  4.5],
       [ 8. ,  2. ,  4. ],
       [15. ,  1. ,  2.5]], dtype=float32)

In [44]:
# Find the rating by our user
x = tiny_ratings_np
the_x_2d = x[np.where(x[:,0] == given_userId)][:, [2]]
the_x_2d = the_x_2d[0].reshape(1,1) # pick the first one in case we have multiple ratings records for user.
the_x = the_x_2d.reshape(the_x_2d.shape[0])
the_x_2d

array([[4.]], dtype=float32)

In [45]:
# Find the ratings by all users
all_x_2d = tiny_ratings_np[:, [2]]
all_x = all_x_2d.reshape(all_x_2d.shape[0])
print (all_x)

all_u_2d = tiny_ratings_np[:, [0]]
all_u = all_u_2d.reshape(all_u_2d.shape[0])

all_m_2d = tiny_ratings_np[:, [1]]
all_m = all_m_2d.reshape(all_m_2d.shape[0])

all_u, all_m, all_x

[4.  4.  4.  4.  5.  3.  4.5 4.  2.5]


(array([ 1.,  1.,  5.,  6.,  6.,  6.,  7.,  8., 15.], dtype=float32),
 array([1., 3., 1., 2., 3., 4., 1., 2., 1.], dtype=float32),
 array([4. , 4. , 4. , 4. , 5. , 3. , 4.5, 4. , 2.5], dtype=float32))

In [46]:
from scipy.spatial.distance import cdist, euclidean
from scipy.spatial import distance_matrix
dm = distance_matrix(all_x_2d, all_x_2d)
dm

array([[0. , 0. , 0. , 0. , 1. , 1. , 0.5, 0. , 1.5],
       [0. , 0. , 0. , 0. , 1. , 1. , 0.5, 0. , 1.5],
       [0. , 0. , 0. , 0. , 1. , 1. , 0.5, 0. , 1.5],
       [0. , 0. , 0. , 0. , 1. , 1. , 0.5, 0. , 1.5],
       [1. , 1. , 1. , 1. , 0. , 2. , 0.5, 1. , 2.5],
       [1. , 1. , 1. , 1. , 2. , 0. , 1.5, 1. , 0.5],
       [0.5, 0.5, 0.5, 0.5, 0.5, 1.5, 0. , 0.5, 2. ],
       [0. , 0. , 0. , 0. , 1. , 1. , 0.5, 0. , 1.5],
       [1.5, 1.5, 1.5, 1.5, 2.5, 0.5, 2. , 1.5, 0. ]], dtype=float32)

In [47]:
# Manually verifying the euclidean calculation.
# The numbers produced by the previous cell and by this cell should match!
dists = cdist(the_x_2d, all_x_2d)
distances = list(dists.reshape(dists.shape[1]))
print (distances)

[0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.5, 0.0, 1.5]


In [48]:
# This cell will throw an exception if the results of euclidean calculation don't match
from math import sqrt
assert(np.isclose(euclidean(the_x, all_x), 
                  sqrt(sum([_*_ for _ in distances]))))

In [49]:
# Cell 13
dists = np.stack([all_u,
                  all_m,
                  np.apply_along_axis(lambda x: sqrt(sum([_*_ for _ in x])), 0, dm)])
dists_df = DataFrame(dists.transpose(), columns=['u', 'm', 'x']) \
           .astype({'u': 'int32', 'm': 'int32', 'x': 'float32'})
dists_df   #  [['u', 'x']]

u  m         x
0   1  1  2.121320
1   1  3  2.121320
2   5  1  2.121320
3   6  2  2.121320
4   6  3  3.937004
5   6  4  3.391165
6   7  1  2.783882
7   8  2  2.121320
8  15  1  4.663690

In [50]:
# Cell 14
# Filtering out the users that are not like our user.
dists_df = dists_df[dists_df['x'] < threshold_distance]
dists_df

u  m         x
0  1  1  2.121320
1  1  3  2.121320
2  5  1  2.121320
3  6  2  2.121320
5  6  4  3.391165
6  7  1  2.783882
7  8  2  2.121320

### Recommending the movies our user may want to watch

In **step 3 of the algorithm**, the system recommends items that the similar users have rated highly but not yet being rated by this user (presumably the absence of rating is often considered as the unfamiliarity of an item).

In [51]:
# Cell 15
# What movies has the user picked already? Don't recommend those!
candidates_df = dists_df[dists_df['u'] == 5]
candidates_np = candidates_df['m']
dists_np = dists_df['m']
candidate_movies = [i for i in set(dists_np) if not i in set(candidates_np) ]
candidate_movies.sort()
print (candidate_movies)

[2, 3, 4]


In [52]:
# Candidate Movies
np.isin(tiny_movies['movieId'], list(candidate_movies))
tiny_movies[np.isin(tiny_movies['movieId'], list(candidate_movies))]

movieId                     title                      genres
1        2            Jumanji (1995)  Adventure|Children|Fantasy
2        3   Grumpier Old Men (1995)              Comedy|Romance
3        4  Waiting to Exhale (1995)        Comedy|Drama|Romance

## TO-DO

Your assignment is to package the code from <span style="color:green"><em># Initial Parameters</em></span> to the <span style="color:green"><em># Candidate Movies</em></span> cells.

1. Create a function `recommend_movies(uid,threshold)` that takes `userId`, `threshold` and `movies_directory` as parameters and produces recommendations for the user. Test the code first with userId = 607. Try various values of threshold such that the user gets at least 6 movie recommendations.
2. Back in the <span style="color:green"><em># What movies has the user picked already? Don't recommend those!</em></span> cell (cell 15), we had picked the first record for our user. Modify the code to begin instead with the movie our user liked the most!
3. Time your code for various values of `userId` and `threshold`. What accounts for the variation in timing?

In [211]:
movies_directory = '../../04-analysis-and-visualization/04-05-recommendations/ml-latest-small/'
from scipy.spatial.distance import cdist, euclidean
from scipy.spatial import distance_matrix
from math import sqrt
import time
def recommend_movies(uid,threshold):

    movies = pd.read_csv(movies_directory+'movies.csv',header = 0)
    ratings = pd.read_csv(movies_directory+'ratings.csv',header = 0).drop(columns=['timestamp'])
    movies_ratings_np = ratings.to_numpy(dtype=np.float32)
    movies_ratings_np =movies_ratings_np[ np.where(movies_ratings_np[:,0] >=600 )]
    # if use all users, it would causes a memory error

    
    
    x = movies_ratings_np
    the_x_2d = x[np.where(x[:,0] == uid)][:, [2]]
    the_x_2d = the_x_2d[0].reshape(1,1) 
    the_x = the_x_2d.reshape(the_x_2d.shape[0])
    


    all_u_2d = movies_ratings_np[:, [0]]
    all_u = all_u_2d.reshape(all_u_2d.shape[0])

    all_m_2d = movies_ratings_np[:, [1]]
    all_m = all_m_2d.reshape(all_m_2d.shape[0])
    
    all_x_2d = movies_ratings_np[:, [2]]
    all_x = all_x_2d.reshape(all_x_2d.shape[0])
    
    dm = distance_matrix(all_x_2d, all_x_2d)
    

    dists = cdist(the_x_2d, all_x_2d)


    distances = list(dists.reshape(dists.shape[1]))
  
    assert(np.isclose(euclidean(the_x, all_x), 
                  sqrt(sum([_*_ for _ in distances]))))
    
    dists = np.stack([all_u,
                  all_m,
                  np.apply_along_axis(lambda x: sqrt(sum([_*_ for _ in x])), 0, dm)])
    dists_df = DataFrame(dists.transpose(), columns=['u', 'm', 'x']) \
           .astype({'u': 'int32', 'm': 'int32', 'x': 'float32'})

#threshold influence ------------------------------------
    dists_df = dists_df[dists_df['x'] < threshold]
    dists_df =dists_df.sort_values(axis = 0,ascending = True,by = 'x')
#threshold influence ------------------------------------




# uid influence ----------------------------------------    other part of influence actually can be ignored
    candidates_df = dists_df[dists_df['u'] == uid]    
    candidates_np = candidates_df['m']
    dists_np = dists_df['m']
    candidate_movies = [i for i in list(dists_np) if not i in set(candidates_np) ]
#uid influence -------------------------------------------- 


    return candidate_movies[0]

start=time.clock() 
recommend_movies(603,200)
end=time.clock()
print(end-start)

1
13.908195999999862


For 1, delete "return candidate_movies[0]" to "return candidate_movies"

For 2, since we use distance to measure, we can simply sort the dists_df by 'x', and then, the "candidate_movies[0]" is the answer.

For 3, changing these 2 value actually has small impact, because it do not influence the calculation of distance which costs most of time. And the "uid" influence the runtime by influencing the part of code that has been tagged (see comments in code) and so does  "threshold"